File: core/logger/logger.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2026 The Coco Sandbox Authors

package logger

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sync"
	"time"
)

// Level represents log level priority
type Level int

const (
	DEBUG Level = iota
	INFO
	WARN
	ERROR
)

func (l Level) String() string {
	switch l {
	case DEBUG:
		return "DEBUG"
	case INFO:
		return "INFO"
	case WARN:
		return "WARN"
	case ERROR:
		return "ERROR"
	default:
		return "UNKNOWN"
	}
}

// Record is a structured log record
type Record struct {
	Timestamp string         `json:"timestamp"`
	Level     string         `json:"level"`
	Component string         `json:"component"`
	Message   string         `json:"message"`
	TraceID   string         `json:"trace_id,omitempty"`
	SandboxID string         `json:"sandbox_id,omitempty"`
	Fields    map[string]any `json:"fields,omitempty"`
}

// Logger provides structured JSON logging
type Logger struct {
	mu        sync.Mutex
	output    io.Writer
	level     Level
	component string
}

// New creates a new Logger for a component
func New(component string) *Logger {
	return &Logger{
		output:    os.Stdout,
		level:     INFO,
		component: component,
	}
}

// SetOutput sets the destination writer
func (l *Logger) SetOutput(w io.Writer) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.output = w
}

// SetLevel sets the minimum log level
func (l *Logger) SetLevel(level Level) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.level = level
}

// WithTraceID returns a logger with trace_id field
func (l *Logger) WithTraceID(traceID string) *Logger {
	return &Logger{
		output:    l.output,
		level:     l.level,
		component: l.component,
	}
}

// WithSandboxID returns a logger with sandbox_id field
func (l *Logger) WithSandboxID(sandboxID string) *Logger {
	return &Logger{
		output:    l.output,
		level:     l.level,
		component: l.component,
	}
}

func (l *Logger) log(level Level, msg string, fields map[string]any) {
	if level < l.level {
		return
	}

	rec := Record{
		Timestamp: time.Now().UTC().Format(time.RFC3339Nano),
		Level:     level.String(),
		Component: l.component,
		Message:   msg,
		Fields:    fields,
	}

	data, err := json.Marshal(rec)
	if err != nil {
		fmt.Fprintf(os.Stderr, "logger: marshal failed: %v\n", err)
		return
	}

	l.mu.Lock()
	defer l.mu.Unlock()
	fmt.Fprintln(l.output, string(data))
}

// Debug logs at DEBUG level
func (l *Logger) Debug(msg string, fields ...map[string]any) {
	f := mergeFields(fields)
	l.log(DEBUG, msg, f)
}

// Info logs at INFO level
func (l *Logger) Info(msg string, fields ...map[string]any) {
	f := mergeFields(fields)
	l.log(INFO, msg, f)
}

// Warn logs at WARN level
func (l *Logger) Warn(msg string, fields ...map[string]any) {
	f := mergeFields(fields)
	l.log(WARN, msg, f)
}

// Error logs at ERROR level
func (l *Logger) Error(msg string, fields ...map[string]any) {
	f := mergeFields(fields)
	l.log(ERROR, msg, f)
}

// Fatal logs at ERROR level and exits
func (l *Logger) Fatal(msg string, fields ...map[string]any) {
	f := mergeFields(fields)
	l.log(ERROR, msg, f)
	os.Exit(1)
}

func mergeFields(fields []map[string]any) map[string]any {
	if len(fields) == 0 {
		return nil
	}
	result := make(map[string]any)
	for _, f := range fields {
		for k, v := range f {
			result[k] = v
		}
	}
	return result
}

// Global default logger for core component
var coreLog = New("coco-core")

// Package-level helpers
func Info(msg string, fields ...map[string]any) { coreLog.Info(msg, fields...) }
func Warn(msg string, fields ...map[string]any) { coreLog.Warn(msg, fields...) }
func Error(msg string, fields ...map[string]any) { coreLog.Error(msg, fields...) }
func Debug(msg string, fields ...map[string]any) { coreLog.Debug(msg, fields...) }

// ForComponent creates a logger for a specific component
func ForComponent(name string) *Logger {
	return New(name)
}

File: daemon/coco-agent/src/logger.zig
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2026 The Coco Sandbox Authors

//! Structured JSON logger for cocod (guest agent).

const std = @import("std");

pub const Level = enum(u8) {
    debug = 0,
    info = 1,
    warn = 2,
    err = 3,

    pub fn toString(self: Level) []const u8 {
        return switch (self) {
            .debug => "DEBUG",
            .info => "INFO",
            .warn => "WARN",
            .err => "ERROR",
        };
    }
};

pub const Record = struct {
    timestamp: []const u8,
    level: []const u8,
    component: []const u8,
    message: []const u8,
    trace_id: ?[]const u8 = null,
    sandbox_id: ?[]const u8 = null,

    pub fn writeJSON(self: Record, writer: anytype) !void {
        try writer.print("{{\"timestamp\":\"{s}\",\"level\":\"{s}\",\"component\":\"{s}\",\"message\":\"", .{
            self.timestamp,
            self.level,
            self.component,
        });

        for (self.message) |c| {
            switch (c) {
                '"' => try writer.writeAll("\\\""),
                '\\' => try writer.writeAll("\\\\"),
                '\n' => try writer.writeAll("\\n"),
                '\r' => try writer.writeAll("\\r"),
                '\t' => try writer.writeAll("\\t"),
                else => try writer.writeByte(c),
            }
        }

        try writer.writeAll("\"");

        if (self.trace_id) |tid| {
            try writer.print(",\"trace_id\":\"{s}\"", .{tid});
        }

        if (self.sandbox_id) |sid| {
            try writer.print(",\"sandbox_id\":\"{s}\"", .{sid});
        }

        try writer.writeAll("}\n");
    }
};

pub const Logger = struct {
    level: Level,
    component: []const u8,
    writer: std.io.Writer(std.fs.File, std.posix.WriteError, std.fs.File.write),

    pub fn init(component: []const u8) Logger {
        return .{
            .level = .info,
            .component = component,
            .writer = std.io.getStdOut().writer(),
        };
    }

    pub fn setLevel(logger: *Logger, level: Level) void {
        logger.level = level;
    }

    fn log(logger: Logger, level: Level, msg: []const u8, trace_id: ?[]const u8, sandbox_id: ?[]const u8) void {
        if (@intFromEnum(level) < @intFromEnum(logger.level)) {
            return;
        }

        var timestamp: [64]u8 = undefined;
        const ts = std.fmt.fmtIntToSlice(timestamp[0..], @as(u64, @intCast(std.time.timestamp())), 10);

        const record = Record{
            .timestamp = ts,
            .level = level.toString(),
            .component = logger.component,
            .message = msg,
            .trace_id = trace_id,
            .sandbox_id = sandbox_id,
        };

        record.writeJSON(logger.writer) catch return;
    }

    pub fn debug(logger: Logger, msg: []const u8) void {
        logger.log(.debug, msg, null, null);
    }

    pub fn info(logger: Logger, msg: []const u8) void {
        logger.log(.info, msg, null, null);
    }

    pub fn warn(logger: Logger, msg: []const u8) void {
        logger.log(.warn, msg, null, null);
    }

    pub fn err(logger: Logger, msg: []const u8) void {
        logger.log(.err, msg, null, null);
    }
};

var global_logger = Logger.init("cocod");

pub fn getLogger() *Logger {
    return &global_logger;
}

pub fn debug(msg: []const u8) void {
    global_logger.log(.debug, msg, null, null);
}
pub fn info(msg: []const u8) void {
    global_logger.log(.info, msg, null, null);
}
pub fn warn(msg: []const u8) void {
    global_logger.log(.warn, msg, null, null);
}
pub fn err(msg: []const u8) void {
    global_logger.log(.err, msg, null, null);
}
